File: pbark/pbark/config/config_loader.py
from __future__ import annotations

from pbark.config._resources import resources_dir
from pbark.config.variable_type import VariableType
from pbark.errors import BarkError
from pbark.lexer import Token, TokenType


class ConfigLoader:
    MEMORY = "memory"
    JOURNAL = "journal"

    _INDEX: dict[str, VariableType] = {}
    _PHRASE_ALIASES: dict[str, str] = {}
    _loaded = False

    @classmethod
    def _ensure_loaded(cls) -> None:
        if cls._loaded:
            return
        cls._load("/breeds.txt", VariableType.BREED)
        cls._load("/objects.txt", VariableType.OBJECT)
        cls._load("/stashes.txt", VariableType.STASH)
        cls._load("/piles.txt", VariableType.PILE)
        cls._INDEX[cls.MEMORY] = VariableType.STORY_NUMBER
        cls._INDEX[cls.JOURNAL] = VariableType.STORY_TEXT
        cls._build_phrase_aliases()
        cls._loaded = True

    @classmethod
    def _load(cls, filename: str, vtype: VariableType) -> None:
        path = resources_dir() / filename.lstrip("/")
        if not path.is_file():
            raise IllegalStateError(f"Resource not found: {filename}")
        for line in path.read_text(encoding="utf-8").splitlines():
            word = line.strip()
            if word:
                cls._INDEX[cls.normalise(word)] = vtype

    @classmethod
    def _build_phrase_aliases(cls) -> None:
        for name in list(cls._INDEX.keys()):
            cls._PHRASE_ALIASES[name] = name
            if "_" in name:
                cls._PHRASE_ALIASES[name.replace("_", " ")] = name

    @staticmethod
    def normalise(name: str) -> str:
        return name.strip().lower()

    @classmethod
    def type_of(cls, name: str) -> VariableType | None:
        cls._ensure_loaded()
        return cls._INDEX.get(cls.normalise(name))

    @classmethod
    def is_valid_name(cls, name: str) -> bool:
        return cls.type_of(name) is not None

    @classmethod
    def is_breed(cls, name: str) -> bool:
        return cls.type_of(name) == VariableType.BREED

    @classmethod
    def is_stash(cls, name: str) -> bool:
        return cls.type_of(name) == VariableType.STASH

    @classmethod
    def is_pile(cls, name: str) -> bool:
        return cls.type_of(name) == VariableType.PILE

    @classmethod
    def is_story_number_constant(cls, name: str) -> bool:
        return cls.type_of(name) == VariableType.STORY_NUMBER

    @classmethod
    def is_story_text_constant(cls, name: str) -> bool:
        return cls.type_of(name) == VariableType.STORY_TEXT

    @classmethod
    def is_story_constant(cls, name: str) -> bool:
        return cls.is_story_number_constant(name)

    @classmethod
    def memory_name(cls) -> str:
        return cls.MEMORY

    @classmethod
    def journal_name(cls) -> str:
        return cls.JOURNAL

    @classmethod
    def token_index_after_resolved_name(
        cls, tokens: list[Token], fr: int, to: int, resolved_name: str
    ) -> int:
        cls._ensure_loaded()
        words: list[str] = []
        positions: list[int] = []
        for i in range(fr, to):
            token = tokens[i]
            if token.is_(TokenType.IDENTIFIER):
                words.append(cls.normalise(token.value))
                positions.append(i)
        for length in range(min(4, len(words)), 0, -1):
            for start in range(len(words) - length + 1):
                phrase = " ".join(words[start : start + length])
                if resolved_name == cls.resolve_name_phrase(phrase):
                    return positions[start + length - 1] + 1
        key = cls.normalise(resolved_name)
        for i in range(fr, to):
            token = tokens[i]
            if token.is_(TokenType.IDENTIFIER) and key == cls.normalise(token.value):
                return i + 1
        return -1

    @classmethod
    def list_names(cls, vtype: VariableType) -> list[str]:
        cls._ensure_loaded()
        return sorted(k for k, v in cls._INDEX.items() if v == vtype)

    @classmethod
    def list_breeds(cls) -> list[str]:
        return cls.list_names(VariableType.BREED)

    @classmethod
    def list_objects(cls) -> list[str]:
        return cls.list_names(VariableType.OBJECT)

    @classmethod
    def list_stashes(cls) -> list[str]:
        return cls.list_names(VariableType.STASH)

    @classmethod
    def list_piles(cls) -> list[str]:
        return cls.list_names(VariableType.PILE)

    @classmethod
    def resolve_name_phrase(cls, phrase: str) -> str | None:
        cls._ensure_loaded()
        return cls._PHRASE_ALIASES.get(cls.normalise(phrase))

    @classmethod
    def resolve_name_from_tokens(cls, tokens: list[Token], fr: int, to: int) -> str | None:
        cls._ensure_loaded()
        words = cls._identifier_words(tokens, fr, to)
        cls._skip_leading_story_noise(words)
        return cls._match_longest_name_phrase(words)

    @classmethod
    def resolve_collection_from_tokens(
        cls, tokens: list[Token], fr: int, to: int
    ) -> str | None:
        cls._ensure_loaded()
        words = cls._identifier_words(tokens, fr, to)
        cls._skip_leading_story_noise(words)
        best = None
        best_length = 0
        for length in range(min(4, len(words)), 0, -1):
            for start in range(len(words) - length + 1):
                phrase = " ".join(words[start : start + length])
                resolved = cls.resolve_name_phrase(phrase)
                if (
                    resolved is not None
                    and (cls.is_stash(resolved) or cls.is_pile(resolved))
                    and length >= best_length
                ):
                    best = resolved
                    best_length = length
        return best

    @classmethod
    def _skip_leading_story_noise(cls, words: list[str]) -> None:
        while words:
            if cls.is_valid_name(words[0]) or cls.resolve_name_phrase(words[0]) is not None:
                break
            name_ahead = False
            for length in range(1, min(4, len(words)) + 1):
                if cls.resolve_name_phrase(" ".join(words[:length])) is not None:
                    name_ahead = True
                    break
            if name_ahead:
                break
            words.pop(0)

    @staticmethod
    def _identifier_words(tokens: list[Token], fr: int, to: int) -> list[str]:
        words: list[str] = []
        for i in range(fr, to):
            token = tokens[i]
            if token.is_(TokenType.IDENTIFIER):
                words.append(ConfigLoader.normalise(token.value))
        return words

    @classmethod
    def _match_longest_name_phrase(cls, words: list[str]) -> str | None:
        for length in range(min(4, len(words)), 0, -1):
            for start in range(len(words) - length + 1):
                phrase = " ".join(words[start : start + length])
                resolved = cls.resolve_name_phrase(phrase)
                if resolved is not None:
                    return resolved
        return None

    @classmethod
    def require_stash(cls, name: str, line: int) -> None:
        if not cls.is_stash(name):
            from pbark.config.name_hints import NameHints

            raise BarkError(
                line,
                f"'{name}' is not a not something the dogs can get stuff from or bury in."
                + NameHints.hint_phrase(name, cls.list_stashes()),
            )

    @classmethod
    def require_pile(cls, name: str, line: int) -> None:
        if not cls.is_pile(name):
            from pbark.config.name_hints import NameHints

            raise BarkError(
                line,
                f"'{name}' is not a registered pile. wrong spot."
                + NameHints.hint_phrase(name, cls.list_piles()),
            )

    @classmethod
    def require_breed(cls, name: str, line: int) -> None:
        if not cls.is_breed(name):
            from pbark.config.name_hints import NameHints

            raise BarkError(
                line,
                f'"{name}" isn\'t a breed we know about. Garbage bin mix??'
                + NameHints.hint_phrase(name, cls.list_breeds()),
            )


class IllegalStateError(RuntimeError):
    pass

File: src/main/java/dev/klomptech/jbark/config/ConfigLoader.java
package dev.klomptech.jbark.config;

import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class ConfigLoader {

  // Global story variables
  public static final String MEMORY = "memory";
  public static final String JOURNAL = "journal";

  private static final Map<String, VariableType> INDEX = new HashMap<>();
  private static final Map<String, String> PHRASE_ALIASES = new HashMap<>();

  private ConfigLoader() {
    // prevent instantiation
  }

  static {
    load("/breeds.txt", VariableType.BREED, INDEX);
    load("/objects.txt", VariableType.OBJECT, INDEX);
    load("/stashes.txt", VariableType.STASH, INDEX);
    load("/piles.txt", VariableType.PILE, INDEX);
    INDEX.put(MEMORY, VariableType.STORY_NUMBER);
    INDEX.put(JOURNAL, VariableType.STORY_TEXT);
    buildPhraseAliases();
  }

  private static void load(
      final String path, final VariableType type, final Map<String, VariableType> index) {
    try (InputStream in = ConfigLoader.class.getResourceAsStream(path)) {
      if (in == null) {
        throw new IllegalStateException("Resource not found: " + path);
      }
      try (BufferedReader reader =
          new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
        reader
            .lines()
            .map(String::trim)
            .filter(line -> !line.isEmpty())
            .map(ConfigLoader::normalise)
            .forEach(
                word -> {
                  index.put(word, type);
                });
      }
    } catch (IOException e) {
      throw new IllegalStateException("Failed to load resource: " + path, e);
    }
  }

  // cookie_jar also answers to "cookie jar" in stories
  private static void buildPhraseAliases() {
    for (String name : INDEX.keySet()) {
      PHRASE_ALIASES.put(name, name);
      if (name.contains("_")) {
        PHRASE_ALIASES.put(name.replace('_', ' '), name);
      }
    }
  }

  public static String normalise(final String name) {
    return name.trim().toLowerCase();
  }

  public static Optional<VariableType> typeOf(final String name) {
    return Optional.ofNullable(INDEX.get(normalise(name)));
  }

  public static boolean isValidName(final String name) {
    return typeOf(name).isPresent();
  }

  public static boolean isBreed(final String name) {
    return typeOf(name).orElse(null) == VariableType.BREED;
  }

  public static boolean isStash(final String name) {
    return typeOf(name).orElse(null) == VariableType.STASH;
  }

  public static boolean isPile(final String name) {
    return typeOf(name).orElse(null) == VariableType.PILE;
  }

  public static boolean isStoryNumberConstant(final String name) {
    return typeOf(name).orElse(null) == VariableType.STORY_NUMBER;
  }

  public static boolean isStoryTextConstant(final String name) {
    return typeOf(name).orElse(null) == VariableType.STORY_TEXT;
  }

  public static boolean isStoryConstant(final String name) {
    return isStoryNumberConstant(name);
  }

  public static String memoryName() {
    return MEMORY;
  }

  public static String journalName() {
    return JOURNAL;
  }

  // Token index after a resolved name phrase on a line slice, or -1
  public static int tokenIndexAfterResolvedName(
      final List<Token> tokens, final int from, final int to, final String resolvedName) {
    List<String> words = new ArrayList<>();
    List<Integer> positions = new ArrayList<>();
    for (int i = from; i < to; i++) {
      Token token = tokens.get(i);
      if (token.is(TokenType.IDENTIFIER)) {
        words.add(normalise(token.value()));
        positions.add(i);
      }
    }
    for (int length = Math.min(4, words.size()); length >= 1; length--) {
      for (int start = 0; start <= words.size() - length; start++) {
        String phrase = String.join(" ", words.subList(start, start + length));
        if (resolvedName.equals(resolveNamePhrase(phrase))) {
          return positions.get(start + length - 1) + 1;
        }
      }
    }
    String key = normalise(resolvedName);
    for (int i = from; i < to; i++) {
      Token token = tokens.get(i);
      if (token.is(TokenType.IDENTIFIER) && key.equals(normalise(token.value()))) {
        return i + 1;
      }
    }
    return -1;
  }

  public static List<String> listNames(final VariableType type) {
    return INDEX.entrySet().stream()
        .filter(entry -> entry.getValue() == type)
        .map(Map.Entry::getKey)
        .sorted()
        .toList();
  }

  public static List<String> listBreeds() {
    return listNames(VariableType.BREED);
  }

  public static List<String> listObjects() {
    return listNames(VariableType.OBJECT);
  }

  public static List<String> listStashes() {
    return listNames(VariableType.STASH);
  }

  public static List<String> listPiles() {
    return listNames(VariableType.PILE);
  }

  public static String resolveNamePhrase(final String phrase) {
    return PHRASE_ALIASES.get(normalise(phrase));
  }

  public static String resolveNameFromTokens(
      final List<Token> tokens, final int from, final int to) {
    List<String> words = identifierWords(tokens, from, to);
    skipLeadingStoryNoise(words);
    return matchLongestNamePhrase(words);
  }

  // Longest registered stash or pile name on a line slice
  // Example: her cookie jar holds "biscuit" -> cookie_jar from stashes.txt.
  public static String resolveCollectionFromTokens(
      final List<Token> tokens, final int from, final int to) {
    List<String> words = identifierWords(tokens, from, to);
    skipLeadingStoryNoise(words);
    String best = null;
    int bestLength = 0;
    for (int length = Math.min(4, words.size()); length >= 1; length--) {
      for (int start = 0; start <= words.size() - length; start++) {
        String phrase = String.join(" ", words.subList(start, start + length));
        String resolved = resolveNamePhrase(phrase);
        if (resolved != null && (isStash(resolved) || isPile(resolved)) && length >= bestLength) {
          best = resolved;
          bestLength = length;
        }
      }
    }
    return best;
  }

  // Drop leading story words (her, my, …) until a registered name phrase can start
  private static void skipLeadingStoryNoise(final List<String> words) {
    while (!words.isEmpty()) {
      if (isValidName(words.get(0)) || resolveNamePhrase(words.get(0)) != null) {
        break;
      }
      boolean nameAhead = false;
      for (int length = 1; length <= Math.min(4, words.size()); length++) {
        if (resolveNamePhrase(String.join(" ", words.subList(0, length))) != null) {
          nameAhead = true;
          break;
        }
      }
      if (nameAhead) {
        break;
      }
      words.remove(0);
    }
  }

  private static List<String> identifierWords(
      final List<Token> tokens, final int from, final int to) {
    List<String> words = new ArrayList<>();
    for (int i = from; i < to; i++) {
      Token token = tokens.get(i);
      if (token.is(TokenType.IDENTIFIER)) {
        words.add(normalise(token.value()));
      }
    }
    return words;
  }

  private static String matchLongestNamePhrase(final List<String> words) {
    for (int length = Math.min(4, words.size()); length >= 1; length--) {
      for (int start = 0; start <= words.size() - length; start++) {
        String phrase = String.join(" ", words.subList(start, start + length));
        String resolved = resolveNamePhrase(phrase);
        if (resolved != null) {
          return resolved;
        }
      }
    }
    return null;
  }

  public static void requireStash(final String name, final int line) throws BarkError {
    if (!isStash(name)) {
      throw new BarkError(
          line,
          "'"
              + name
              + "' is not a not something the dogs can get stuff from or bury in."
              + NameHints.hintPhrase(name, listStashes()));
    }
  }

  public static void requirePile(final String name, final int line) throws BarkError {
    if (!isPile(name)) {
      throw new BarkError(
          line,
          "'"
              + name
              + "' is not a registered pile. wrong spot."
              + NameHints.hintPhrase(name, listPiles()));
    }
  }

  public static void requireBreed(final String name, final int line) throws BarkError {
    if (!isBreed(name)) {
      throw new BarkError(
          line,
          "\""
              + name
              + "\" isn't a breed we know about. Garbage bin mix??"
              + NameHints.hintPhrase(name, listBreeds()));
    }
  }
}
